scripts/xtensa-build-all.sh

```diff
@@ -8,87 +8,64 @@ BUILD_ROM=no
 BUILD_DEBUG=no
 BUILD_FORCE_UP=no
 BUILD_JOBS=$(nproc --all)
-BUILD_JOBS_NEXT=0
 BUILD_VERBOSE=
 
 PATH=$pwd/local/bin:$PATH
 
-pwd=`pwd`
+pwd=$(pwd)
+
+print_usage()
+{
+    cat <<EOF
+usage: xtensa-build.sh [options] platform(s)
+       [-r] Build rom (gcc only)
+       [-a] Build all platforms
+       [-u] Force UP ARCH
+       [-d] Enable debug build
+       [-c] Configure defconfig
+       [-k] Use private key
+       [-v] Verbose Makefile log
+       [-j n] Set number of make build jobs. Jobs=#cores when no flag. \
+Inifinte when not specified.
+       Supported platforms ${SUPPORTED_PLATFORMS[@]}
+EOF
+}
+
+# parse the args
+while getopts "rudj:ckva" OPTION; do
+        case "$OPTION" in
+		r) BUILD_ROM=yes ;;
+		u) BUILD_FORCE_UP=yes ;;
+		d) BUILD_DEBUG=yes ;;
+		j) BUILD_JOBS=$OPTARG ;;
+		c) MAKE_MENUCONFIG=yes ;;
+		k) USE_PRIVATE_KEY=yes ;;
+		v) BUILD_VERBOSE='VERBOSE=1' ;;
+		a) PLATFORMS=${SUPPORTED_PLATFORMS[@]} ;;
+		*) print_usage; exit 1 ;;
+        esac
+done
+shift $((OPTIND-1))
 
-if [ "$#" -eq 0 ]
-then
-	echo "usage: xtensa-build.sh [options] platform(s)"
-	echo "       [-r] Build rom (gcc only)"
-	echo "       [-a] Build all platforms"
-	echo "       [-u] Force UP ARCH"
-	echo "       [-d] Enable debug build"
-	echo "       [-c] Configure defconfig"
-	echo "       [-k] Use private key"
-	echo "       [-v] Verbose Makefile log"
-	echo "       [-j [n]] Set number of make build jobs." \
-		"Jobs=#cores when no flag. Inifinte when no arg."
-	echo "       Supported platforms ${SUPPORTED_PLATFORMS[@]}"
-else
-	# parse the args
-	for args in $@
+# parse platform args
+for arg in "$@"; do
+	platform=none
+	for i in "${SUPPORTED_PLATFORMS[@]}"
 	do
-		if [[ "$args" == "-r" ]]
-			then
-			BUILD_ROM=yes
-
-		elif [[ "$args" == "-u" ]]
-			then
-			BUILD_FORCE_UP=yes
-
-		elif [[ "$args" == "-d" ]]
-			then
-			BUILD_DEBUG=yes
-
-		elif [[ "$args" == "-j" ]]
-			then
-			BUILD_JOBS_NEXT=1
-			BUILD_JOBS=""
-
-		elif [[ "$args" == "-c" ]]
-			then
-			MAKE_MENUCONFIG=yes
-
-		elif [[ "$args" == "-k" ]]
-			then
-			USE_PRIVATE_KEY=yes
-
-		elif [[ "$args" == "-v" ]]
-			then
-			BUILD_VERBOSE='VERBOSE=1'
-
-		# Build all platforms
-		elif [[ "$args" == "-a" ]]
-			then
-			PLATFORMS=${SUPPORTED_PLATFORMS[@]}
-		else
-			# check for plaform
-			for i in ${SUPPORTED_PLATFORMS[@]}
-			do
-				if [ $i == $args ]
-				then
-					PLATFORMS+=$i" "
-					BUILD_JOBS_NEXT=0
-					continue 2
-				fi
-			done
-
-			# check for jobs
-			if [ ${BUILD_JOBS_NEXT} == 1 ]
-				then
-				BUILD_JOBS=$args
-				BUILD_JOBS_NEXT=0
-				continue
-			fi
-
-			printf '%s: WARN: ignoring arg %s\n' "$0" "$args" >&2
+		if [ $i == $arg ]
+		then
+			PLATFORMS+=$i" "
```

Comment:
That's not how you expand arrays, I'm surprised this works...

https://mywiki.wooledge.org/BashGuide/Arrays

Reply:
Wow, I got to the bottom of this PLATFORMS thing and it's pretty bad. It's technically unrelated to the changes you're doing but I'm unfortunately not OK refactoring and moving around code that bad, even when you're not making it worse. Whoever wrote `PLATFORMS=${SUPPORTED_PLATFORMS[@]}` expected `PLATFORMS` to be an array. It's not, it's just one  string with spaces. This is ironic considering the main purpose of bash arrays is to manage list of strings with spaces in each; I digress.

Of course `PLATFORMS+=$i" "` is even "less" an array in this `else` clause. Maybe the author didn't really understand that arrays and strings are different types.

Yet there are references to PLATFORMS[@] below that seem to work [by coincidence](https://pragprog.com/the-pragmatic-programmer/extracts/coincidence).

Reply:
BTW I just saw the same array/string confusion in `./scripts/qemu-check.sh`